File: Algoexpert/First Duplicate Value/cpp/first_duplicate_value.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    public:
        int firstDuplicateValueHashSet(vector<int> array) {
            unordered_map<int, bool> seen;
            for(int value : array) {
                if (seen.find(value) != seen.end()) {
                    return value;
                }
                seen[value] = true;
            }
            return -1;
        }

        int firstDuplicateValueOptimized(vector<int> array) {
            for (int value : array) {
                int absValue = abs(value);
                if( array[absValue-1] < 0) {
                    return absValue;
                }
                array[absValue-1] *= -1;
            }
            return -1;
        }
};


int main() {

    vector<int> array = {2, 1, 5, 2, 3, 3, 4};

    Solution solution;
    cout <<  solution.firstDuplicateValueHashSet(array) << endl;
    cout <<  solution.firstDuplicateOptimized(array) << endl;

    return 0;
}

File: Algoexpert/First Duplicate Value/java/src/FirstDuplicateValue.java
import java.util.HashSet;
import java.util.Set;

public class FirstDuplicateValue {

    public static int firstDuplicateValueHashSet(int[] array) {
        Set<Integer> found = new HashSet<>();
        for(int i = 0; i < array.length - 1; i++) {
            if (found.contains(array[i])) return array[i];
            found.add(array[i]);
        }
        return -1;
    }

    public static int firstDuplicateValueOptimized(int[] array) {
        for (int value : array) {
            int absValue = Math.abs(value);
            if( array[absValue-1] < 0) {
                return absValue;
            }
            array[absValue-1] *= -1;
        }
        return -1;
    }

    public static void main(String[] args) {

        int[] array = {2, 1, 5, 2, 3, 3, 4};

        System.out.println(firstDuplicateValueHashSet(array));
        System.out.println(firstDuplicateValueOptimized(array));
    }

}
